lib/asset-sprayer/src/prompts.rs

```diff
@@ -0,0 +1,147 @@
+
+#[derive(Debug, Clone)]
+pub struct Prompts {
+    prompts_dir: Option<PathBuf>,
+}
+
+#[derive(Debug, Clone, Copy, strum::Display)]
+pub enum Prompt {
+    AssetSchema,
+}
+
+impl Prompt {
+    fn yaml_relative_path(&self) -> &str {
+        match self {
+            Prompt::AssetSchema => "aws/asset_schema.yaml",
+        }
+    }
+
+    fn yaml_embedded(&self) -> &'static str {
+        match self {
+            Prompt::AssetSchema => include_str!("../prompts/aws/asset_schema.yaml"),
+        }
+    }
+}
+
+impl Prompts {
+    pub fn new(prompts_dir: Option<PathBuf>) -> Self {
+        Self {
+            prompts_dir: prompts_dir.map(Into::into),
+        }
+    }
+
+    pub async fn create_request(
+        &self,
+        prompt: Prompt,
+        replace: &[(&str, &str)],
+    ) -> AssetSprayerResult<CreateChatCompletionRequest> {
+        let request = self.raw_request(prompt).await?;
+        Self::replace_prompt_request(request, replace, prompt).await
+    }
+
+    async fn raw_request(&self, prompt: Prompt) -> AssetSprayerResult<CreateChatCompletionRequest> {
+        Ok(serde_yaml::from_str(&self.yaml(prompt).await?)?)
+    }
+
+    async fn yaml(&self, prompt: Prompt) -> AssetSprayerResult<String> {
+        if let Some(ref prompts_dir) = self.prompts_dir {
+            // Read from disk if prompts_dir is available (faster dev cycle)
+            let path = prompts_dir.join(prompt.yaml_relative_path());
+            Ok(tokio::fs::read_to_string(path).await?)
+        } else {
+            Ok(prompt.yaml_embedded().to_string())
+        }
+    }
+
+    async fn replace_prompt_request(
+        request: CreateChatCompletionRequest,
+        replace: &[(&str, &str)],
+        prompt: Prompt,
+    ) -> AssetSprayerResult<CreateChatCompletionRequest> {
+        let mut request = request;
+        for message in request.messages.iter_mut() {
+            *message =
+                Self::replace_prompt_request_message(message.clone(), replace, prompt).await?;
+        }
+        Ok(request)
+    }
+
+    async fn replace_prompt_request_message(
+        message: ChatCompletionRequestMessage,
+        replace: &[(&str, &str)],
+        prompt: Prompt,
+    ) -> AssetSprayerResult<ChatCompletionRequestMessage> {
+        let mut message = message;
+        match &mut message {
+            ChatCompletionRequestMessage::User(message) => {
+                if let ChatCompletionRequestUserMessageContent::Text(text) = &mut message.content {
+                    *text = Self::replace_prompt_text(text.clone(), replace, prompt).await?;
+                }
+            }
+            ChatCompletionRequestMessage::System(message) => {
+                if let ChatCompletionRequestSystemMessageContent::Text(text) = &mut message.content
+                {
+                    *text = Self::replace_prompt_text(text.clone(), replace, prompt).await?;
+                }
+            }
+            _ => (),
+        }
+        Ok(message)
+    }
```

Comment:
Is mutating the message passed in not a desirable option?

Reply:
In some places it was better to return a new thing, and some places &mut was better, but since they were all the same function stack I just wanted to pick a lane :)